CS328-Notes/notebooks/2022_04_10_DistinctCount_BloomFilters.ipynb

```diff
@@ -0,0 +1,304 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "21c5d52b",
+   "metadata": {},
+   "source": [
+    "# Bloom Filters\n",
+    "\n",
+    "### Querying\n",
+    "\n",
+    "Suppose we have a large number of elements in our universe U. This could be anything ranging from books in a library to IP addresses stored in a server. Now suppose we want to implement a data structure that would help us in querying whether a particular element is present in the set or not, efficiently in terms of time and space. Since we have much lesser memory as compared to the size of our universe, we have to find a way to reduce the memory requirement for the querying data structure from $|U|$ to something less. Similarly, we can’t spend $O(|U|)$ time(a linear search) for querying as this would be too much. So, what are the possible solutions?\n",
+    "The first idea would be to use Hashing; which reduces the space used to $O(n log |U|)$ and the querying time to $O(n/m)$(where $n$ is the number of elements stored and $m$ is the size of the hash table), or to use a bit array which uses $O(|U|)$ space but queries in $O(1)$ time on average. But let us go beyond this and try and figure out if we can do better! Turns out, if we allow errors in querying with a small probability delta(a Monte Carlo algorithm), we can, using a data structure known as a Bloom Filter. \n",
+    "\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "b212957a",
+   "metadata": {},
+   "source": [
+    "### Bloom Filter:"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "id": "f19fac4f",
+   "metadata": {},
+   "source": [
+    "```{figure} ../assets/2022_04_10_DistinctCount_BloomFilters/fig1.png\n",
+    "---\n",
+    "name: Figure 1: Bloom Filter\n",
```

Comment:
name, i guess, should be like a reference/variable name (like `bloom_filter_fig`, `fig72378`, `bloom-fig`)
Put "Bloom Filter" in the caption (not name)